File: 백준/Silver/1018. 체스판 다시 칠하기/체스판 다시 칠하기.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	// Main
	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		map = new char[N][M];

		for (int i = 0; i < N; i++) {
			String str = br.readLine();
			for (int j = 0; j < M; j++) {
				map[i][j] = str.charAt(j);
			}
		}

		int min = Integer.MAX_VALUE;
		for (int i = 0; i <= N - 8; i++) {
			for (int j = 0; j <= M - 8; j++) {
				min = Math.min(min, whatIsFirst('W', i, j));
				min = Math.min(min, whatIsFirst('B', i, j));
			}
		}
		System.out.println(min);

	}

	static char[][] map;
	static int N;
	static int M;

	private static int whatIsFirst(char color, int r, int c) {
		int change = 0;
		for (int i = r; i < r + 8; i++) {
			for (int j = c; j < c + 8; j++) {
				if ((i + j) % 2 == 0 && map[i][j] != color) {
					change++;
				}
				if ((i + j) % 2 == 1 && map[i][j] == color) {
					change++;
				}
			}
		}

		return change;
	}

}

File: 백준/Silver/1018. 체스판 다시 칠하기/체스판 다시 칠하기.py
n, m = map(int, input().split())

grid = [list(input()) for i in range(n)]
# print(grid)
max = float('inf')


def changeColor(r, c, color):
    change = 0
    for i in range(r, r + 8):
        for j in range(c, c + 8):
            if ((i + j) % 2 == 0 and grid[i][j] != color):
                change += 1
            if ((i + j) % 2 == 1 and grid[i][j] == color):
                change += 1

    return change


for i in range(0, n - 8 + 1):
    for j in range(0, m - 8 + 1):
        # i,j 는 시작점임
        max = min(max, changeColor(i, j, "W"))
        max = min(max, changeColor(i, j, "B"))

print(max)
